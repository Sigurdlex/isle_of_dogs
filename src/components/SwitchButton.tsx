import React, { ReactNode } from 'react';
import styled from 'styled-components';

interface SwitchButtonProps {
  handleLeftClick: () => void;
  leftColor: string;
  leftContent: ReactNode;
  handleRightClick: () => void;
  rightColor: string;
  rightContent: ReactNode;
  isLeft: boolean;
}
interface ButtonProps {
  color: string;
  left?: boolean;
}

export const SwitchButton = ({
  handleLeftClick,
  leftColor,
  leftContent,
  handleRightClick,
  rightColor,
  rightContent,
  isLeft,
}: SwitchButtonProps) => (
  <div>
    <Button
      onClick={handleLeftClick}
      color={leftColor}
      left
      disabled={isLeft}
      data-testid="left-button"
    >
      {leftContent}
    </Button>
    <Button
      onClick={handleRightClick}
      color={rightColor}
      disabled={!isLeft}
      data-testid="right-button"
    >
      {rightContent}
    </Button>
  </div>
);

const Button = styled.button<ButtonProps>`
  background: #e5e5e5;
  color: rgba(0, 0, 0, 0.5);
  font-size: inherit;
  padding: ${({ left }) => (left ? '.5em 1em .5em 2em' : '.5em 2em .5em 1em')};
  width: 120px;
  border: none;
  border-radius: ${({ left }) => (left ? '1em 0 0 1em' : '0 1em 1em 0')};
  position: relative;
  transition: 0.2s all;
  transform-origin: ${({ left }) => (left ? '100%' : 0)} 50%;
  &:disabled {
    background: white;
    color: black;
    cursor: default;
    transform: scale(1.05);
  }
  &::before {
    width: 6px;
    height: 100%;
    ${({ left }) => (left ? 'left' : 'right')}: 1em;
    position: absolute;
    background: ${({ color }) => color};
    content: '';
    top: 0;
  }
`;
