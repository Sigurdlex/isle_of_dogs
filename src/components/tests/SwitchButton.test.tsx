import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import 'jest-styled-components';

import { SwitchButton } from '../SwitchButton';

const handleLeftClick = jest.fn();
const handleRightClick = jest.fn();
const leftContent = 'leftContent';
const rightContent = 'rightContent';
const leftColor = 'red';
const rightColor = 'green';

afterEach(() => {
  cleanup();
  handleLeftClick.mockClear();
  handleRightClick.mockClear();
});

test('<SwitchButton /> left', () => {
  const { getByTestId, container } = render(
    <SwitchButton
      handleLeftClick={handleLeftClick}
      leftColor={leftColor}
      leftContent={leftContent}
      handleRightClick={handleRightClick}
      rightColor={rightColor}
      rightContent={rightContent}
      isLeft
    />
  );
  const element = container.firstChild as HTMLElement;
  const leftButton: HTMLButtonElement = getByTestId('left-button') as HTMLButtonElement;
  const rightButton: HTMLButtonElement = getByTestId('right-button') as HTMLButtonElement;
  expect(leftButton.disabled).toBeTruthy();
  expect(rightButton.disabled).toBeFalsy();
  fireEvent.click(rightButton);
  expect(handleRightClick).toHaveBeenCalledTimes(1);
  fireEvent.click(leftButton);
  expect(handleLeftClick).toHaveBeenCalledTimes(0);
  expect(leftButton.textContent).toMatch(leftContent);
  expect(leftButton).toHaveStyleRule('padding', '.5em 1em .5em 2em');
  expect(leftButton).toHaveStyleRule('border-radius', '1em 0 0 1em');
  expect(leftButton).toHaveStyleRule('transform-origin', '100% 50%');
  expect(handleRightClick).toHaveBeenCalledTimes(1);
  expect(rightButton.textContent).toMatch(rightContent);
  expect(rightButton).toHaveStyleRule('padding', '.5em 2em .5em 1em');
  expect(rightButton).toHaveStyleRule('border-radius', '0 1em 1em 0');
  expect(rightButton).toHaveStyleRule('transform-origin', '0 50%');
  expect(element).toMatchSnapshot();
});

test('<SwitchButton /> right', () => {
  const { getByTestId, container: { firstChild: container } } = render(
    <SwitchButton
      handleLeftClick={handleLeftClick}
      leftColor={leftColor}
      leftContent={leftContent}
      handleRightClick={handleRightClick}
      rightColor={rightColor}
      rightContent={rightContent}
      isLeft={false}
    />
  );
  const leftButton: HTMLButtonElement = getByTestId('left-button') as HTMLButtonElement;
  const rightButton: HTMLButtonElement = getByTestId('right-button') as HTMLButtonElement;
  expect(leftButton.disabled).toBeFalsy();
  expect(rightButton.disabled).toBeTruthy();
  fireEvent.click(rightButton);
  expect(handleRightClick).toHaveBeenCalledTimes(0);
  fireEvent.click(leftButton);
  expect(handleLeftClick).toHaveBeenCalledTimes(1);
  expect(container).toMatchSnapshot();
});
