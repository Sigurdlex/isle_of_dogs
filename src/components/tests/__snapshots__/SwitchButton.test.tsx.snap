// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SwitchButton /> left 1`] = `
.c0 {
  background: #e5e5e5;
  color: rgba(0,0,0,0.5);
  font-size: inherit;
  padding: .5em 1em .5em 2em;
  width: 120px;
  border: none;
  border-radius: 1em 0 0 1em;
  position: relative;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  -webkit-transform-origin: 100% 50%;
  -ms-transform-origin: 100% 50%;
  transform-origin: 100% 50%;
}

.c0:disabled {
  background: white;
  color: black;
  cursor: default;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c0::before {
  width: 6px;
  height: 100%;
  left: 1em;
  position: absolute;
  background: red;
  content: '';
  top: 0;
}

.c1 {
  background: #e5e5e5;
  color: rgba(0,0,0,0.5);
  font-size: inherit;
  padding: .5em 2em .5em 1em;
  width: 120px;
  border: none;
  border-radius: 0 1em 1em 0;
  position: relative;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  -webkit-transform-origin: 0 50%;
  -ms-transform-origin: 0 50%;
  transform-origin: 0 50%;
}

.c1:disabled {
  background: white;
  color: black;
  cursor: default;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1::before {
  width: 6px;
  height: 100%;
  right: 1em;
  position: absolute;
  background: green;
  content: '';
  top: 0;
}

<div>
  <button
    class="c0"
    color="red"
    data-testid="left-button"
    disabled=""
  >
    leftContent
  </button>
  <button
    class="c1"
    color="green"
    data-testid="right-button"
  >
    rightContent
  </button>
</div>
`;

exports[`<SwitchButton /> right 1`] = `
.c0 {
  background: #e5e5e5;
  color: rgba(0,0,0,0.5);
  font-size: inherit;
  padding: .5em 1em .5em 2em;
  width: 120px;
  border: none;
  border-radius: 1em 0 0 1em;
  position: relative;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  -webkit-transform-origin: 100% 50%;
  -ms-transform-origin: 100% 50%;
  transform-origin: 100% 50%;
}

.c0:disabled {
  background: white;
  color: black;
  cursor: default;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c0::before {
  width: 6px;
  height: 100%;
  left: 1em;
  position: absolute;
  background: red;
  content: '';
  top: 0;
}

.c1 {
  background: #e5e5e5;
  color: rgba(0,0,0,0.5);
  font-size: inherit;
  padding: .5em 2em .5em 1em;
  width: 120px;
  border: none;
  border-radius: 0 1em 1em 0;
  position: relative;
  -webkit-transition: 0.2s all;
  transition: 0.2s all;
  -webkit-transform-origin: 0 50%;
  -ms-transform-origin: 0 50%;
  transform-origin: 0 50%;
}

.c1:disabled {
  background: white;
  color: black;
  cursor: default;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1::before {
  width: 6px;
  height: 100%;
  right: 1em;
  position: absolute;
  background: green;
  content: '';
  top: 0;
}

<div>
  <button
    class="c0"
    color="red"
    data-testid="left-button"
  >
    leftContent
  </button>
  <button
    class="c1"
    color="green"
    data-testid="right-button"
    disabled=""
  >
    rightContent
  </button>
</div>
`;
